### PR TITLE
loop-util: attach an empty file before unlocking device

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/shared/loop-util.c [hunk 1]
```diff
@@ -145,16 +145,17 @@ static int loop_configure(
          * block devices. */
         r = device_has_block_children(nr);
         if (r < 0)
-                return r;
+                return log_debug_errno(r, "Failed to determine if the loopback device has partitions: %m");
         if (r > 0) {
                 r = loop_is_bound(fd);
                 if (r < 0)
-                        return r;
+                        return log_debug_errno(r, "Failed to determine if a file descritor is attached to the loopback device: %m");
```
Comment (mrc0mmand) [inline src/shared/loop-util.c]: descritor -> descriptor

**Review by mrc0mmand — APPROVED** [review]
The loopback devices are truly an infinite source of _fun_.

**Comment by bluca** [comment]
Why is the diskseq stuff being removed? We got it added to newer kernels to fix these races - of course won't help with older kernels, but at some point if should work out no?

**Comment by yuwata** [comment]
> Why is the diskseq stuff being removed? We got it added to newer kernels to fix these races - of course won't help with older kernels, but at some point if should work out no?

As you can see, at least now it is not used, ie the argument is completely ignored. Please revert it when it is used.

**Comment by yuwata** [comment]
Or, are there any places I dropped in this PR where the diskseq is really used?

**Comment by bluca** [comment]
But it was used when it was introduced? Where did that stuff go?

**Comment by yuwata** [comment]
I cannot dig the git history now from cellphone, but I guess these entries in LoopDevice were stopped to be used when Lennart stop to use sd_device_monitor for detecting loopback devices.

**Comment by bluca** [comment]
@poettering do you remember what's the full story? It was quite hard to get the diskseq stuff in the kernel and it was explicitly to fix these issues, it's a pity if it was all for naught...

**Comment by yuwata** [comment]
@bluca diskseq or friends are not used since 1b010ae7c96d953f11a9925d4c5aef29befc86d1.

**Comment by yuwata** [comment]
BTW @mrc0mmand Does this help issue #24450?? Could you run a stress test for TEST-50-DISSECT with this PR?

**Comment by mrc0mmand** [comment]
> BTW @mrc0mmand Does this help issue #24450?? Could you run a stress test for TEST-50-DISSECT with this PR?

I ran TEST-50-DISSECT 100 times in a row with this PR and it didn't fail a single time, so I'm inclined to say it might potentially fix that issue (but won't say it directly, otherwise something else breaks).

File: src/shared/loop-util.h [hunk 2]
```diff
@@ -14,9 +14,6 @@ struct LoopDevice {
         dev_t devno;
         char *node;
         bool relinquished;
-        uint64_t diskseq; /* Block device sequence number, monothonically incremented by the kernel on create/attach, or 0 if we don't know */
```
Comment (poettering) [inline src/shared/loop-util.h:17]: I'd really keep the diskseq, it's how we can pin our use of the loopback device. Even if the current dissection stuff doesn#t use it, it probably should, and condition the partition removal to only execute if it still matches the original diskseq (enclosed in a BSD lock)

File: src/shared/loop-util.c [hunk 3]
```diff
@@ -55,16 +55,23 @@ static int loop_is_bound(int fd) {
         return true; /* bound! */
 }
 
-static int device_has_block_children(sd_device *d) {
+static int device_has_block_children(int nr) {
```
Comment (poettering) [inline src/shared/loop-util.c:58]: should be renamed to `loop_block_device_has_children()` or so. It doesn't work for generic block devices anymore, after all, just for loopback devices

File: src/shared/loop-util.c [hunk 4]
```diff
@@ -109,6 +109,44 @@ static int device_has_block_children(int nr) {
...
+static int attach_empty_file(int loop) {
+        _cleanup_close_ int fd = -1;
+
+        /* So here's the thing: on various kernels (5.8 at least) loop block devices might enter a state
+         * where they are detached but nonetheless have partitions, when used heavily. Accessing these
+         * partitions results in immediatey IO errors. There's no pretty way to get rid of them
+         * again. Neither LOOP_CLR_FD nor LOOP_CTL_REMOVE suffice (see above). What does work is to
+         * reassociate them with a new fd however. This is what we do here hence: we associate the devices
+         * with an empty file (i.e. an image that definitely has no partitions). We then immediately clear it
+         * again. This suffices to make the partitions go away. Ugly but appears to work. */
```
Comment (poettering) [inline src/shared/loop-util.c:121]: so, back when i added this code i wasn't aware of BLKPG_DEL_PARTITION. I wonder if a loop around that wouldn't have the same effect but be less ugly?

i.e. remove all partitions in a loop, until none are left?

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
Inspired by this I prepped a PR in #24530, which fixes a bunch of related things. It also gets rid of the attach empty file logic entirely, in favour of explicit partition deletion, as I suggested above.

That new PR makes the last two of the commits in this PR redundant, I guess. (one because it drops the empty file attach stuff, the other because we actually add proper locking around clear+delete of block devices. at least in my tests this makes things entirely clean in test-loop-block even with 200 threads and 200 iterations)

**Comment by yuwata** [comment]
@poettering Thanks. I will update or close this after your PR is merged.

**Pull request closed without merging**